transmart-core-db/src/main/groovy/org/transmartproject/db/multidimquery/query/HibernateCriteriaQueryBuilder.groovy

```diff
@@ -237,101 +240,149 @@ class HibernateCriteriaQueryBuilder implements QueryBuilder<Criterion, DetachedC
     }
 
     /**
-     * Creates a criteria object for a field constraint. Applies {@link #convertValue(Field, Object)} on the value
+     * Creates a {@link Criterion} for the Boolean constraint that operates
+     * on a property and a value.
+     *
      * Supports the operators:
-     * - EQUALS
-     * - GREATER_THAN
-     * - GREATER_THAN_OR_EQUALS
-     * - LESS_THAN
-     * - LESS_THAN_OR_EQUALS
-     * - CONTAINS (both for collections and strings)
-     * - LIKE
-     * @throws QueryBuilderException if the field type does not support the operator or the value is not supported
-     * for the field type.
-     * @see {@link Operator} and {@link Type} for supported operators and types.
+     * - {@link Operator#EQUALS}
+     * - {@link Operator#NOT_EQUALS}
+     * - {@link Operator#GREATER_THAN}
+     * - {@link Operator#GREATER_THAN_OR_EQUALS}
+     * - {@link Operator#LESS_THAN}
+     * - {@link Operator#LESS_THAN_OR_EQUALS}
+     * - {@link Operator#BEFORE}
+     * - {@link Operator#AFTER}
+     * - {@link Operator#BETWEEN}
+     * - {@link Operator#CONTAINS} (both for collections and strings)
+     * - {@link Operator#LIKE}
+     * - {@link Operator#IN}
+     *
+     * @param operator the operator to apply
+     * @param propertyName the name of the property used as left hand side of the operation
+     * @param type the type of the property
+     * @param value the value used as right hand side of the operation
+     * @return a {@link Criterion} object represention the operation.
      */
-    Criterion build(FieldConstraint constraint) {
-        assert constraint.field != null
-        if (!constraint.operator.supportsType(constraint.field.type)) {
-            throw new QueryBuilderException("Field type ${constraint.field.type} not supported for operator '${constraint.operator.symbol}'.")
-        }
-        if ((!(constraint.operator in [Operator.BETWEEN, Operator.IN]) && !constraint.field.type.supportsValue(constraint.value))
-                || (constraint.operator in [Operator.BETWEEN, Operator.IN] && !constraint.field.type.supportsValue(constraint.value.first()))) {
-            throw new QueryBuilderException("Value of class ${constraint.value?.class?.simpleName} not supported for field type '${constraint.field.type}'.")
-        }
-        constraint.value = convertValue(constraint.field, constraint.value)
-        String propertyName = getFieldPropertyName(constraint.field)
-        switch(constraint.operator) {
+    static Criterion applyOperator(Operator operator, String propertyName, Type type, Object value) {
+        Criterion criterion
+        switch(operator) {
             case Operator.EQUALS:
-                return Restrictions.eq(propertyName, constraint.value)
+                criterion = Restrictions.eq(propertyName, value)
+                break
```

Comment:
This looks like a significant uglification with all the `break`s now needed. How about wrapping the switch in its own method so you can keep using return?